### PR TITLE
Prevent recursion error and allow PhysicalType objects to be pickled

### LINKED ISSUE #11685: PhysicalType instances are not pickle-able
### Description

In `main`, `PhysicalType` instances are not pickle-able with dev numpy.

### Steps to Reproduce

```python
import pickle
import astropy.units as u

phys_type = u.get_physical_type('length')

with open('/tmp/test.pkl', 'wb') as f:
    pickle.dump(phys_type, f)
    
with open('/tmp/test.pkl', 'rb') as f:
    pickle.load(f)
```

### Exception

```python
---------------------------------------------------------------------------
RecursionError                            Traceback (most recent call last)
<ipython-input-14-7ceef6c385b0> in <module>
      5 
      6 with open('/tmp/test.pkl', 'rb') as f:
----> 7     pickle.load(f)

~/projects/astropy-all/astropy/astropy/units/physical.py in __getattr__(self, attr)
    315 
    316     def __getattr__(self, attr):
--> 317         self_str_attr = getattr(str(self), attr, None)
    318         if hasattr(str(self), attr):
    319             warning_message = (

~/projects/astropy-all/astropy/astropy/units/physical.py in __str__(self)
    354 
    355     def __str__(self):
--> 356         return "/".join(self._physical_type_list)
    357 
    358     @staticmethod

... last 2 frames repeated, from the frame below ...

~/projects/astropy-all/astropy/astropy/units/physical.py in __getattr__(self, attr)
    315 
    316     def __getattr__(self, attr):
--> 317         self_str_attr = getattr(str(self), attr, None)
    318         if hasattr(str(self), attr):
    319             warning_message = (

RecursionError: maximum recursion depth exceeded
```

### System Details

Tested on macOS 10.15.7 and ubuntu 18.04

Seems to only happen with numpy 1.21.0.dev, which I don't fully understand...

cc @namurphy @mhvk

### PR DESCRIPTION
**Pull request by adrn**
Fixes #11685

### CODE REVIEW

File: astropy/units/tests/test_physical.py [hunk 1]
```diff
@@ -503,3 +504,15 @@ def test_missing_physical_type_attribute():
...
+@pytest.mark.parametrize('ptype', [
+    u.get_physical_type('length'),
+    u.get_physical_type('speed'),
+    u.get_physical_type('entropy')
+])
+def test_pickling(ptype):
+    # Regression test for #11685
+    pkl = pickle.dumps(ptype)
+    other = pickle.loads(pkl)
+    assert other == ptype
```
Comment (namurphy) [inline astropy/units/tests/test_physical.py:518]:
```suggestion
@pytest.mark.parametrize('ptype_name', ['length', 'speed', 'entropy'])
def test_pickling(ptype_name):
    # Regression test for #11685
    ptype = u.get_physical_type(ptype_name)
    pkl = pickle.dumps(ptype)
    other = pickle.loads(pkl)
    assert other == ptype
```
Thank you for adding this test!  In retrospect, I should have added this in #11204.  Here's an optional suggestion that could make it a tad more straightforward to parse.

File: astropy/units/physical.py [hunk 2]
```diff
@@ -314,6 +314,13 @@ def __iter__(self):
         yield from self._physical_type_list
 
     def __getattr__(self, attr):
+        # TODO: remove this whole method when accessing str attributes from
+        # physical types is no longer supported
+
+        # short circuit attribute accessed in __str__ to prevent recursion
+        if attr == '_physical_type_list':
+            super().__getattribute__(attr)
```
Comment (namurphy) [inline astropy/units/physical.py:322]: Interesting!  I was pretty confused by this error, but I'm glad this fixes it.  Kind thanks!

**Review by namurphy — APPROVED** [review]
Thank you for the quick fix!  I was pretty confused about why the `RecursionError` was happening, but that probably means I ought to read the Python docs on `__getattr__` and `__getattribute__`.  I have one optional suggestion and it overall looks good.

File: astropy/units/tests/test_physical.py (hunk 1 above)
Comment (adrn) [reply astropy/units/tests/test_physical.py:518]: Oh yea - that looks better :)

**Review by pllim — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**